src/reader.rs

```diff
@@ -76,25 +76,26 @@ impl Ext4Read for Vec<u8> {
         start_byte: u64,
         dst: &mut [u8],
     ) -> Result<(), BoxedError> {
-        let start = usize::try_from(start_byte).map_err(|_| {
+        read_from_bytes(self, start_byte, dst).ok_or_else(|| {
             Box::new(MemIoError {
                 start: start_byte,
                 read_len: dst.len(),
                 src_len: self.len(),
             })
-        })?;
-
-        let end = start + dst.len();
-        let src = self.get(start..end).ok_or(Box::new(MemIoError {
-            start: start_byte,
-            read_len: dst.len(),
-            src_len: self.len(),
-        }))?;
-        dst.copy_from_slice(src);
-        Ok(())
+            .into()
+        })
     }
 }
 
+fn read_from_bytes(src: &[u8], start_byte: u64, dst: &mut [u8]) -> Option<()> {
```

Comment:
The separate function is really just for the convenience of being able to use `?`. All the error paths end up with the same error, and this way we don't have to repeat that code or call `.map_err`/`ok_or_else` more than once.

Rust has an unstable feature called try blocks that would be a nicer solution to this problem: https://doc.rust-lang.org/beta/unstable-book/language-features/try-blocks.html. This has been in progress since 2016 so who knows if it'll ever be stabilized.

Reply:
I got the point of the separate function, I was asking why this function doesn't return a `Result`. I think `Result` is what I would've reached for here, so I was trying to figure out if you see something I don't.

Reply:
With a `Result` I couldn't use `?`, without the verbosity of `map_err`/`ok_or`/`ok_or_else`. So really just saving a bit of typing.